lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp

```diff
@@ -46,16 +46,20 @@ static void appendPossiblyAbsolutePath(SmallVectorImpl<char> &base,
   }
 }
 
+namespace llvm {
+/// Hashing for StringAttr.
+template <>
+struct DenseMapInfo<StringAttr> : public DenseMapInfo<Attribute> {
+  static StringAttr getEmptyKey() { return {}; }
+  static StringAttr getTombstoneKey() { return {}; }
```

Comment:
Nice, I didn't know that you could just return the same key as empty and tombstone key. That's pretty handy. Inheriting from `Attribute`'s `DenseMapInfo` is very elegant -- did not think of that :+1:

Reply:
Uh yeah, good noticing that.  I don't think you can use the same value for the tombstone and I'll fix it.

Reply:
I copied some code from the Attribute DenseMapInfo, it uses the default pointer DenseMapInfo to create a Attribute::Impl and constructs a StringAttr from that.  Its kinda reaching in to the implementation of Attributes too much for my liking, so I'm not sure what to do.  Any idea if there is a sane way to create these?

Reply:
If I remember correctly, implementing `PointerLikeTypeTraits` on `StringAttr` should work, and then just forwarding to the pointer traits impl of the underlying `Attribute`. There's a blanket impl of `DenseMapInfo` on `PointerLikeTypeTraits` somewhere such that you should be able to use anything pointer like in a `DenseMap` automatically.

Reply:
Ended up switching to using a `SmallPtrSet<Attribute>`.